### PR TITLE
[DC-2062] Transition write_results_to_(drc|site)_bucket to use buckets

### PR DESCRIPTION
**Pull request by MichaelSchmidt82**
Scope of changes:

**Functions:**

1. Modified:

data_steward/validation/participants/writers/ create_site_validation_report(...)

- The function accepts a Blob instead of a bucket string.  The blob uses GCS to provide direct upload functionality.  
- A new try and expect guards against failure.  Error and success logging are still present.
- The function accepts a Client instead of the project string.  Otherwise, there are two instances of a patched client or a mocked client in tests without this change.  In addition, the change makes testing possible utilizing a singleton pattern.
- Deriving the project string is now possible from the client parameter via client.project attribute.
- The function returns a single value for simplification.  Previously the first returned value was never used, and the second returned value was used only in tests.
- I added type hinting and followed advice from pylint.

2. Modified:
data_steward/validation/participants/identity_match/write_results_to_site_buckets()

Reason for change: This function calls the modified create_site_validation_report function above.

- The function accepts a client following the singleton pattern for testing.  Before the change, it received a string whose sole purpose was passing it down.
- The function utilizes changes in DC-2063 -- using buckets from get_hpo_bucket()
- A Blob is created and passed down instead of a filename.
- Added a try and except guard for bucket acquisition and blob creation.  Failures are now logged.

3. Modified:
data_steward/validation/participants/identity_match/write_results_to_drc_bucket(...)

Reason for change: This function calls the modified create_site_validation_report function above.

- The function accepts a client following the singleton pattern for testing.  Before the change, it received a string whose sole purpose was passing it down.
- The function utilizes changes in DC-2063 -- using buckets from get_drc_bucket()
- A Blob is created and passed down instead of a filename.
- Added a try and except guard for bucket acquisition and blob creation.  Failures are now logged.

**Tests:**

1. Modified
tests/unit_tests/data_steward/validation/participants/writer_tests.WriterTest.test_create_site_validation_report(...)

- Fitted a mocked Blob to capture the attributes used (required).
- A mocked client provides the project string (required).
- Blob.upload_from_file(...) is checked against call count and call parameters (not required, but follows best practice).
- I applied type hinting and followed the advice from pylint.

2. Modified
tests/unit_tests/data_steward/validation/participants/writer_tests.WriterTest.test_create_site_validation_report_with_errors(...)

- I applied the analogous changes from the non-error test version. 

3. Modified
tests/unit_tests/data_steward/validation/participants/identity_match_test.IdentityMatchTest.setUp

- I renamed some of the variable values for easier identification.  For example, 'aou-awe2' was renamed 'fake_bucket_3' (because it is the third element in a list).
- I patched the bucket returning methods get_drc_bucket() and get_hpo_bukect(...) from the client.  The changes have the same effect as the previous string returning methods.

4. Modified
tests/unit_tests/data_steward/validation/participants/identity_match_test.IdentityMatchTest.test_write_results_to_site_buckets()

- Changes only reflect the use of clients, buckets, and blobs.  (required)
- I removed the filename variable and applied the corresponding value to a blob's name field.  (required)
- It's worth noting that mocked clients' return mocked buckets and have mocked blobs.  So the required functionality found in GCS is copied.
- I applied type hints and followed recommendations from pylint.

5. Modified
tests/unit_tests/data_steward/validation/participants/identity_match_test.IdentityMatchTest.test_write_results_to_site_buckets_simulate_errors()

- I applied the analogous changes found in the non-error version.

tests/unit_tests/data_steward/validation/participants/identity_match_test.IdentityMatchTest.test_write_results_to_drc_bucket()

- I applied the analogous changes from the non-error site function.
- This function has fewer modifications as it uses a single blob instead of multiple ones.

tests/unit_tests/data_steward/validation/participants/identity_match_test.IdentityMatchTest.test_write_results_to_drc_bucket_simulate_errors()

- I applied the analogous changes from the error site function.
- This function has fewer modifications as it uses a single blob instead of multiple ones.

### CODE REVIEW

File: data_steward/gcloud/gcs/__init__.py
Comment (nishanthpp93) [inline data_steward/gcloud/gcs/__init__.py:65]: This should be `hpo_id`. It is the identifier used for HPO sites in the project.

File: data_steward/gcloud/gcs/__init__.py
Comment (nishanthpp93) [inline data_steward/gcloud/gcs/__init__.py:80]: Same here about `hpo_id`

File: data_steward/gcloud/gcs/__init__.py
Comment (nishanthpp93) [inline data_steward/gcloud/gcs/__init__.py:127]: I'd suggest `_get_hpo_bucket_name` since you're using the string `bucket_name` in usages above.
```suggestion
    def _get_hpo_bucket_name(self, hpo_id: str) -> str:
```

**Review by nishanthpp93 — CHANGES_REQUESTED** [review]
Only glanced at `gcs/__init__.py` and added a couple comments. I'd suggest other reviewers to chime in as well.

File: data_steward/gcloud/gcs/__init__.py
Comment (MichaelSchmidt82) [reply data_steward/gcloud/gcs/__init__.py]: Committed

File: data_steward/gcloud/gcs/__init__.py
Comment (MichaelSchmidt82) [reply data_steward/gcloud/gcs/__init__.py]: Committed

File: data_steward/gcloud/gcs/__init__.py
Comment (MichaelSchmidt82) [reply data_steward/gcloud/gcs/__init__.py:127]: Committed.  I agree these are better names.

**Comment by MichaelSchmidt82** [comment]
> Only glanced at `gcs/__init__.py` and added a couple comments. I'd suggest other reviewers to chime in as well.

It's also worth noting that some values and variable renames make the diff messy.  The core changes are listed in the PR,  but I'm going to add some comments at code line numbers for clarity.

File: data_steward/validation/main.py
Comment (MichaelSchmidt82) [inline data_steward/validation/main.py:895]: This exception handling was added for reliablity.

File: data_steward/validation/main.py
Comment (MichaelSchmidt82) [inline data_steward/validation/main.py:929]: This "added" code is actually some of the deletions above, but the addition of exception handling and the movement of other exceptions caused this code to shift.

File: data_steward/validation/participants/identity_match.py
Comment (MichaelSchmidt82) [inline data_steward/validation/participants/identity_match.py:646]: Full use of StorageClient.

File: data_steward/validation/participants/identity_match.py
Comment (MichaelSchmidt82) [inline data_steward/validation/participants/identity_match.py:622]: Singleton pattern for client.

File: data_steward/validation/participants/identity_match.py
Comment (MichaelSchmidt82) [inline data_steward/validation/participants/identity_match.py:683]: Removal of strings in place of buckets was added here.  Exception handling added that previously didn't exitst.

File: data_steward/gcloud/gcs/__init__.py
Comment (MichaelSchmidt82) [inline data_steward/gcloud/gcs/__init__.py:127]: Help function provides functionality for @hiro-mishima DC-1817

File: data_steward/validation/participants/writers.py
Comment (MichaelSchmidt82) [inline data_steward/validation/participants/writers.py:70]: Direct use of buckets and blobs.

File: data_steward/validation/participants/writers.py
Comment (MichaelSchmidt82) [inline data_steward/validation/participants/writers.py:235]: Usage of `client.project`, `bucket.name` and `blob.name` when strings are required.

File: tests/unit_tests/data_steward/validation/main_test.py
Comment (MichaelSchmidt82) [inline tests/unit_tests/data_steward/validation/main_test.py:474]: Sections here replicate GCS client, bucket, and blob creation.

File: tests/unit_tests/data_steward/validation/participants/identity_match_test.py
Comment (MichaelSchmidt82) [inline tests/unit_tests/data_steward/validation/participants/identity_match_test.py:28]: Variable and value renames

File: tests/unit_tests/data_steward/validation/participants/identity_match_test.py
Comment (MichaelSchmidt82) [inline tests/unit_tests/data_steward/validation/participants/identity_match_test.py:497]: Use of client, not string

File: tests/unit_tests/data_steward/validation/participants/identity_match_test.py
Comment (MichaelSchmidt82) [inline tests/unit_tests/data_steward/validation/participants/identity_match_test.py:473]: Use of client, not string.

File: tests/unit_tests/data_steward/validation/participants/writers_test.py
Comment (MichaelSchmidt82) [inline tests/unit_tests/data_steward/validation/participants/writers_test.py:37]: Changes here reflect the use of buckets.

File: data_steward/validation/main.py
Comment (lledezma) [inline data_steward/validation/main.py:947]: I would delete the if statement in line 945 since allowed_items is a filtered list coming from bucket_items in line 908.

File: tests/unit_tests/data_steward/validation/main_test.py
Comment (lledezma) [inline tests/unit_tests/data_steward/validation/main_test.py:475]: Repeated pre-conditions already defined in lines 458 and 459.

**Review by lledezma — CHANGES_REQUESTED** [review]

File: data_steward/validation/main.py
Comment (MichaelSchmidt82) [reply data_steward/validation/main.py:947]: Yeah.  I can see that.  I think I reasoned that if you capture a blob that doesn't exist, you will get something like "NoneType doesn't copy_blob() method."   Bucket's that grab blobs that don't exist, "fail" silently, and return None. But you're correct; you have a guarantee.  I'll shuffle some around and remove the if block.

File: tests/unit_tests/data_steward/validation/main_test.py
Comment (MichaelSchmidt82) [reply tests/unit_tests/data_steward/validation/main_test.py:475]: Sometimes you get something working and forget to clean up your mess. Committed.

**Review by lledezma — DISMISSED** [review]

File: data_steward/validation/main.py
Comment (MichaelSchmidt82) [reply data_steward/validation/main.py]: Refactored in 2063 per your advice and brought in

File: data_steward/gcloud/gcs/__init__.py
Comment (nishanthpp93) [inline data_steward/gcloud/gcs/__init__.py:102]: Nice, we can change this once we get approval from Sysadmin.

File: data_steward/validation/participants/identity_match.py
Comment (nishanthpp93) [inline data_steward/validation/participants/identity_match.py:635]: +1 for Pythonic

File: data_steward/validation/participants/identity_match.py
Comment (nishanthpp93) [inline data_steward/validation/participants/identity_match.py:650]: Nice, this ensures other sites can continue to be written to if one fails

File: data_steward/validation/participants/writers.py
Comment (nishanthpp93) [inline data_steward/validation/participants/writers.py:10]: These should go under third party imports

**Review by nishanthpp93 — APPROVED** [review]
Added a suggestion for imports, but other than that looks good and the tests pass!

File: data_steward/validation/participants/identity_match.py
Comment (MichaelSchmidt82) [reply data_steward/validation/participants/identity_match.py:650]: I had to double-check that we can use `exc.message` instead of `exc`, so this won't raise while raising --  `exc.message` won't raise an attribute error.

File: data_steward/validation/participants/writers.py
Comment (MichaelSchmidt82) [reply data_steward/validation/participants/writers.py:10]: Good catch -- got to check my linter more often

File: data_steward/validation/participants/identity_match.py
Comment (MichaelSchmidt82) [reply data_steward/validation/participants/identity_match.py:635]: The real question is which ones get printed:

```
print('using is')
if '' is None:
    print("'' is none")

if [] is None:
    print('[] is None')

print('using not')
if not []:
    print('not []')

if not '':
    print("not ''")

print("we're done here")
```

File: data_steward/validation/participants/identity_match.py
Comment (nishanthpp93) [reply data_steward/validation/participants/identity_match.py:650]: Oh right, this is something that I discussed with @lledezma as well for BucketNotSet. Adding a message attribute helped with custom exceptions. Google Cloud seems to have it already set for their exceptions.

**Pull request merged**
